Support `minimize_energy!` for systems with vacancies

Minimal reproducer:

```jl
using Sunny

latvecs = lattice_vectors(1, 1, 10, 90, 90, 90)
cryst = Crystal(latvecs, [[0,0,0]])
sys = System(cryst, (10,10,1), [SpinInfo(1, S=1, g=2)], :dipole; seed=1)
sys2 = to_inhomogeneous(sys)

set_vacancy_at!(sys2, (1,1,1,1))
minimize_energy!(sys2) # AssertionError: n' * n ≈ 1
```

Bug reported by Lázaro Cabán Cabrera.

--- a/docs/src/versions.md
+++ b/docs/src/versions.md
@@ -4,6 +4,7 @@
 (In development)
 
 * Better error message when a $g$-tensor is symmetry disallowed.
+* Fix [`minimize_energy!`](@ref) when used with [`set_vacancy_at!`](@ref).
 
 ## v0.7.3
 (Nov 12, 2024)

--- a/src/Optimization.jl
+++ b/src/Optimization.jl
@@ -4,7 +4,7 @@
 # normalized. When `α=0`, the output is `u=n`, and when `|α|→ ∞` the output is
 # `u=-n`. In all cases, `|u|=1`.
 function stereographic_projection(α, n)
-    @assert n'*n ≈ 1 || n != n
+    @assert n'*n ≈ 1 || all(isnan, n)
 
     v = α - n*(n'*α)              # project out component parallel to `n`
     v² = real(v'*v)
@@ -31,7 +31,7 @@ end
 #   x̄ du/dα = x̄ du/dv P
 #
 @inline function vjp_stereographic_projection(x̄, α, n)
-    n != n && return zero(Vec3) # No gradient when α = [0, 0, 0] is fixed
+    all(isnan, n) && return zero(Vec3) # No gradient when α = [0, 0, 0] is fixed
 
     @assert n'*n ≈ 1
 
@@ -47,7 +47,7 @@ end
 
 # Returns v such that u = (2v + (1-v²)n)/(1+v²) and v⋅n = 0
 function inverse_stereographic_projection(u, n)
-    n != n && return zero(Vec3) # NaN values in n denote α = v = [0, 0, 0]
+    all(isnan, n) && return zero(Vec3) # NaN values denote α = v = [0, 0, 0]
 
     @assert u'*u ≈ 1
 

--- a/src/Optimization.jl
+++ b/src/Optimization.jl
@@ -4,7 +4,7 @@
 # normalized. When `α=0`, the output is `u=n`, and when `|α|→ ∞` the output is
 # `u=-n`. In all cases, `|u|=1`.
 function stereographic_projection(α, n)
-    @assert n'*n ≈ 1
+    @assert n'*n ≈ 1 || n != n
 
     v = α - n*(n'*α)              # project out component parallel to `n`
     v² = real(v'*v)
@@ -30,7 +30,9 @@ end
 # Using the above definitions, return:
 #   x̄ du/dα = x̄ du/dv P
 #
-@inline function vjp_stereographic_projection(x, α, n)
+@inline function vjp_stereographic_projection(x̄, α, n)
+    n != n && return zero(Vec3) # No gradient when α = [0, 0, 0] is fixed
+
     @assert n'*n ≈ 1
 
     v = α - n*(n'*α)
@@ -38,13 +40,15 @@ end
     b = (1-v²)/2
     c = 2/(1+v²)
     # Perform dot products first to avoid constructing outer-product
-    x̄_dudv = c*x' - c * (x' * ((1+c*b)*n + c*v)) * v'
+    x̄_dudv = c*x̄' - c * (x̄' * ((1+c*b)*n + c*v)) * v'
     # Apply projection P=1-nn̄ on right
     return x̄_dudv - (x̄_dudv * n) * n'
 end
 
 # Returns v such that u = (2v + (1-v²)n)/(1+v²) and v⋅n = 0
 function inverse_stereographic_projection(u, n)
+    n != n && return zero(Vec3) # NaN values in n denote α = v = [0, 0, 0]
+
     @assert u'*u ≈ 1
 
     uperp = u - n*(n'*u)

--- a/test/test_energy_consistency.jl
+++ b/test/test_energy_consistency.jl
@@ -65,7 +65,7 @@
             # Pick a random site, try to set it to a random spin
             site = rand(sys.rng, eachsite(sys))
             spin = Sunny.randspin(sys, site)
-            
+
             ΔE = Sunny.local_energy_change(sys, site, spin)
 
             E0 = energy(sys)

--- a/test/test_optimization.jl
+++ b/test/test_optimization.jl
@@ -88,3 +88,22 @@ end
     @test minimize_energy!(sys_dip) > 0
     @test minimize_energy!(sys_sun) > 0
 end
+
+
+@testitem "Vacancies" begin
+    latvecs = lattice_vectors(1, 1, 10, 90, 90, 90)
+    cryst = Crystal(latvecs, [[0,0,0]])
+    sys = System(cryst, [1 => Moment(s=1, g=2)], :dipole; dims=(3, 3, 1), seed=1)
+    set_exchange!(sys, 1, Bond(1, 1, [1, 0, 0]))
+    sys2 = to_inhomogeneous(sys)
+    sys2.κs[1, 1, 1, 1] = 0.01
+    randomize_spins!(sys2)
+    minimize_energy!(sys2)
+    energy_per_site(sys2)
+    @test energy_per_site(sys2) ≈ -0.8889
+    
+    set_vacancy_at!(sys2, (1, 1, 1, 1))
+    randomize_spins!(sys2)
+    minimize_energy!(sys2)
+    @test energy_per_site(sys2) ≈ -8/9
+end

